### ISSUE TITLE
Smarter list widget additions

### ISSUE THREAD

**Issue by erquhart** [netlify/netlify-cms#1244]
**- Do you want to request a *feature* or report a *bug*?**

enhancement

**- What is the current behavior?**

Adding items to the list widget is painful - you have to:

* scroll to the add button up top
* click it
* scroll to the bottom of the list
* edit the newly added item
* reposition the item to desired location

**- What is the expected behavior?**

* scroll to desired new item location
* click an in-location add item button
* edit the new item
* rejoice

**Comment by tech4him1** [comment]
Related to https://github.com/netlify/netlify-cms/issues/963

**Comment by erquhart** [comment]
Agreed, this can be considered a sub-ticket under #963.

**Comment by siliconchild** [comment]
**possible solutions** 

1. Add the 'add new list item' button to the bottom part of the list as well as the top to avoid rescrolling'
2. Alternative solution is to add a 'scroll to the top of the list' button at the bottom of the list widget

NB - Displaying list items like how blog posts are displayed, like this - https://nupic.co/image/6TnWX would free up a lot of vertical space

**Comment by mittalyashu** [comment]
Very simple design, hope you like it 😋

![group 2](https://user-images.githubusercontent.com/29014463/44978224-cbd54e80-af87-11e8-91ea-51e89386fd92.png)

Here's the [source file](https://www.figma.com/file/0bTxEnrHCNM32hDPDQJvrzbr/NetlifyCMS-Design?node-id=96%3A28).

**Comment by erquhart** [comment]
Interesting! What's the red dot for?

**Comment by mittalyashu** [comment]
There is still a room for improvement, it's just that I wasn't sure how people will write their `config.yml` file. 😅

> What's the red dot for?

https://youtu.be/l28MCjo-Jxs?t=581

**Comment by austincondiff** [comment]
I completely agree that this is very painful. I don't know why this isn't here already, especially considering the amount of vertical space each list item takes.

**Comment by erquhart** [comment]
This is open source, not a paid service, so the community does a fair amount of the heavy lifting. If you're up to contribute I'm happy to answer questions and help any way I can.

**Comment by melbourne2991** [comment]
I just put this together - code is pretty much ready to go - thoughts?

![Peek 2019-08-30 12-05](https://user-images.githubusercontent.com/4619280/63988291-ed2b3500-cb1e-11e9-987c-e76f809d1f6c.gif)

**Comment by erquhart** [comment]
Thanks for digging into this @melbourne2991! I still think the approach I mentioned would be ideal - the jumping is totally avoidable with CSS. If you can update to what you had I can help with that part.

**Comment by melbourne2991** [comment]
@erquhart I pulled some of the code out into this sandbox to help illustrate the various approaches:
https://codesandbox.io/s/solitary-rgb-brg22

There are two visual styles and two "strategies" for displaying the insert button:

Samples A and B always show the button based on the mouse position within the list.
C and D only display the button when hovering _between_ the items.

A is closest to what you initially described I believe? C and D feel kind of jarring to me, but maybe that's just a matter preference. I like B as I find it to be least intrusive but I can see how A might be more intuitive/obvious as it clearly highlights where the item will be inserted.

Should I push on with A?

**Comment by ZoltanVeres** [comment]
This is excellent work, truly awesome that we test out versions!
My bet is on A, because it visualizes where the item will go, B could also work too. 
C and D is not going to work on mobile, because it only uses the hover state so can’t add items from mobile.
I know the CMS is not fully optimized for mobile, but we should avoid solutions which exclude these usages.

**Comment by tomrutgers** [comment]
IMHO B is easiest to use. Maybe the button should be slightly more subtle if it's always going to be there. I'm also curious about the flow with the types list widget! BTW Homerun.co does something similar, I think it works really well: https://demo.homerun.co/?demo=job-editor

**Comment by austincondiff** [comment]
I really think there is too much movement in all of these examples. There may be a larger problem to be solved here. 

Adding items in between two others seems to be a use-case easily solvable by dragging a newly created item to the desired place. It seems overly redundant to have the button appear in between. 

If we wanted this functionality, each item might have a menu where we could add two items "Insert one above" and "Insert one below". While we are at it we might add a "Duplicate" option as well (in my experience this would be very beneficial). This way things are out of your way unless you need it.

**Comment by austincondiff** [comment]
Okay, here is a super rough mockup of what I am talking about in the above comment. Didn't have a lot of time but you get the idea. 

![image](https://user-images.githubusercontent.com/806104/64459303-2df5f180-d0bd-11e9-8fa9-1c824907de0f.png)

**Comment by melbourne2991** [comment]
@austincondiff I like this. It has the additional benefit of being able to work on mobile flawlessly. Do we need the move up / move down buttons though? I think rearranging is much easier with drag and drop, which is the current behavior.

**Comment by erquhart** [comment]
Inserting between items is important. You wouldn't think so until you actually need to position items regularly in a long list (some have hundreds). WordPress Gutenberg uses buttons that appear between items on hover and it works great, though they're definitely not the first.

@melbourne2991 thanks for putting this together, I'm on mobile right now but I'll check it out next time I'm at my machine.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by austincondiff** [comment]
@melbourne2991 @erquhart Yeah, I figured adding items in between was important, especially when each list item has many fields. In this concept we would still be able to drag and drop. We don't necessarily need these menu items, however I would imagine it improves usability on mobile where it may be difficult for some to drag and drop. Some mobile browser experiences make this pretty difficult depending on the device.

**Comment by erquhart** [comment]
Example of an indicator that appears in the space between blocks without any impact to surrounding elements: https://wordpress.org/gutenberg/

That page is the live Gutenberg editor, try inserting a block between blocks. We could probably use something thin and wide since the space between list items has no other purpose.

**Comment by austincondiff** [comment]
@erquhart It is interesting because it looks like Gutenberg does both... 
![image](https://user-images.githubusercontent.com/806104/64567316-980cd180-d31d-11e9-8e91-ca8fc6e5f0d3.png)

I will say that this is a page builder use-case and not a repetitive list as in our situation, but there might be room to draw inspiration here.

**Comment by austincondiff** [comment]
Under the design outlined in #2557, this would work well...
![image](https://user-images.githubusercontent.com/806104/64568267-ee7b0f80-d31f-11e9-88e5-0106fc79ef8d.png)

Hover state...
![image](https://user-images.githubusercontent.com/806104/64568430-57fb1e00-d320-11e9-97ff-5f3985edaef6.png)

If the list has types...
![image](https://user-images.githubusercontent.com/806104/64568707-0bfca900-d321-11e9-8f4c-180557126013.png)

**Comment by erquhart** [comment]
Come to think of it, between the two approaches, context menu is the only one that would work on a touch device, so maybe it is the best initial approach.

**Comment by austincondiff** [comment]
@erquhart I've made a sandbox to demonstrate both of these concepts

https://codesandbox.io/embed/netlify-cms-editor-design-j1cg3

As mentioned https://github.com/netlify/netlify-cms/issues/2557#issuecomment-532785406 it's not perfect but it should give you an idea as to how this might work.

**Comment by erquhart** [comment]
That sandbox is pretty fly, nice work! For this immediate feature being something we can implement without expanding scope, I think adding just a blue line that shows on hover in the existing space between items is probably our best bet, whereas your proof of concept would likely be a part of a larger rework - at a minimum its a rework of the list widget. Would love to see some of that added to the current list widget 👍👍

**Comment by erquhart** [comment]
Using Google Slides and noticed there's no direct way to insert a new slide, but if a slide is selected, the new slide will be inserted after the selected slide. This works because the slides panel scrolls independently of the rest of the UI, and the add slide button is in a fixed toolbar.

The more I consider this approach the more it seems extremely ideal, not just for smarter list item insertions, but for improving the UX for nested widgets like object and list. The markdown widget toolbar is sticky, so for long markdown widgets, the toolbar sticks to the top of the screen until you pass the end of the widget. What if we did this for the object and list widgets? The toolbar is always visible, which also means you can collapse the whole thing at any point. For multiple levels of nested widgets, the sticky toolbars can stack, and hierarchy will be obvious. We could even make the toolbar thinner when it's sticking.

Thoughts?
